Arm backend: Int16 linear support

per opened the pull request:
### Summary
Adds support for a16w8 for linear when targeting a backend with +int16 extension.

Fixes #13729

### Test plan
Tested through unit tests.

cc @digantdesai @freddan80 @zingo @oscarandersson8218

per committed:
Arm backend: Decompose conv2d with 16 bit activation

Support quantization to 16a8w. Since the resulting TOSA operator
needs to have the bias in int48 which isn't avaiable as a type in
torch, the conv2d needs to be decomposed into a conv + add, where
the conv result is scaled down to 32 bit before the addition of the
bias is done.

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: Ib8cae694035796374a55a9909e501596e983abf5

per committed:
Arm backend: Handle i48 special case for bias tensor

For the case when the activation is 16 bit the bias in TOSA must be
a int48_t tensor. Since that can't be represented using torch.dtypes
the corresponding node.meta is set with a key 'tosa_dtype_48bit' to
pass through the note to the creation of the TOSA Tensor.
Also make sure to distinguish between int32 and int48 tensors in fuse
constant ops pass.

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: Iefe64f2b02f388c905c9c818ee7d2a6af40bc9e3

per committed:
Arm backend: Handle 16 bit activation for conv2d

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: I2b189b559f699c7eda6921ed515c0e8a849226ca

per committed:
Arm backend: Fix mult and scale calculation for int48_t

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: Ibe158d8d35a632547290f1b9a055d061ae267d77

per committed:
Arm backend: Enable linear 16a8w tests

Enable tests of int16 activations and int8 weight quantization.
Test for large_rand is disabled to sort out why the test is flaky.

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: I9de5d472f8862edebcf82c140399985db930c069

per commented:
Unrelated failures.

digantdesai commented on backends/arm/_passes/add_bias_pass.py:
```diff
@@ -55,6 +55,8 @@ def call(self, graph_module):
                         name=f"{node.name}_bias",
                     )
+                    if node.args[0].meta["val"].dtype == torch.int16:
+                        bias_node.meta["tosa_dtype_48bit"] = True
```
Nit:
```suggestion
                        bias_node.meta["tosa_dtype"] = tosa_dtype.INT48
```

digantdesai commented on backends/arm/_passes/arm_pass_manager.py:
```diff
@@ -167,9 +169,14 @@ def _tosa_INT_pipeline(self, exported_program: ExportedProgram) -> GraphModule:
         self.add_pass(FuseConstantArgsPass(exported_program))
+        self.add_pass(InsertTableOpsPass(exported_program))
+        # If we have a conv2d with int16 activation split up into a convolution
+        # and an addition, to work-around the lack of support for int48 in torch
```
> # and an addition, to work-around the lack of support for int48 in torch
Can it be done by using torch.dtype.int64 instead and then detecting and lowering it as int48 downstream?

digantdesai reviewed (approved):
This is awesome, No conv tests for 16a8w, just curious.

per committed:
Arm backend: Add special dtype TOSA handling

Add a enum class to handle special dtypes that can't be represented
in torch (i.e. int48_t) to avoid leaking serializer types into the
pass handling of the backend.

Signed-off-by: Per Åstrand <per.astrand@arm.com>
Change-Id: I3388cec3c8a26f28790eedc3f124c336b6724cb4

digantdesai commented on backends/arm/test/ops/test_linear.py:
```diff
@@ -276,10 +275,19 @@ def get_symmetric_a16w8_linear_quantizer(
 )
+
+
+@common.parametrize("test_data", test_data_all_16a8w)
```
@per - Does it work with U55 and U85?
@Ninja91 - IIRC you already have tests seems like haven't merged?

per replied:
No, not yet. Test will be added for Ethos-U with a Vela update when it's in place.

zingo commented:
I see this testfail :( 
[test-arm-backend (test_pytest_ops_ethosu_fvp) / linux-job https://github.com/pytorch/executorch/actions/runs/17768877335/job/50517971935?pr=14258#logs

FAILED backends/arm/test/ops/test_linear.py::test_linear_16a8w_tosa_INT[model_linear_rank1_large_randn,per_channel_quant=True] - AssertionError: Output 0 does not match reference output.
	Given atol: 0.004528127912431955, rtol: 0.001.
	Output tensor shape: torch.Size([20]), dtype: torch.float32
	Difference: max: 0.1481781005859375, abs: 0.1481781005859375, mean abs error: 0.008467483520507812.
	-- Model vs. Reference --
	 Numel: 20, 20
	Median: 14.398289680480957, 14.398289680480957
	  Mean: 2.563008487224579, 2.555246722698212
	   Max: 69.8498764038086, 69.84635162353516
	   Min: -115.46151733398438, -115.60969543457031

zingo commented:
I removed milestone 1.0 for now Vela does not support this anyway so lets take this a bit calmer.

digantdesai commented:
yeah let me try to merge this.

zingo commented:
Fails are unrelated
Note: Fail in trunk / test-arm-ootb-linux / linux-job  is unrelated and handled by another PR

digantdesai committed:
Merge branch 'main' into int16_linear_support
